vendor/brave-ios/Ledger/Data/BATLedgerDatabase.mm

```diff
@@ -514,6 +560,49 @@ + (void)insertPendingContributions:(NSArray<BATPendingContribution *> *)contribu
   } completion:WriteToDataControllerCompletion(completion)];
 }
 
++ (void)removePendingContributionForPublisherID:(NSString *)publisherID viewingID:(NSString *)viewingID addedDate:(UInt64)addedDate completion:(BATLedgerDatabaseWriteCompletion)completion
+{
+  // Early guard to check if object exists.
+  // The check happens on main thread, while the deletion is done in background.
+  const auto pendingContribution = [self getPendingContributonWithPublisherID:publisherID
+                                                                    viewingID:viewingID
+                                                                    addedDate:addedDate
+                                                                      context:DataController.viewContext];
+  if (!pendingContribution) {
+    if (completion) { completion(NO); }
+    return;
+  }
+  
+  [DataController.shared performOnContext:nil task:^(NSManagedObjectContext * _Nonnull context) {
+    const auto pendingContribution = [self getPendingContributonWithPublisherID:publisherID
+                                                                      viewingID:viewingID
+                                                                      addedDate:addedDate
+                                                                        context:context];
+    [context deleteObject:pendingContribution];
+  } completion:WriteToDataControllerCompletion(completion)];
+}
+
++ (void)removeAllPendingContributions:(BATLedgerDatabaseWriteCompletion)completion
+{
+  [DataController.shared performOnContext:nil task:^(NSManagedObjectContext * _Nonnull context) {
+    const auto fetchRequest = PendingContribution.fetchRequest;
+    fetchRequest.entity = [NSEntityDescription entityForName:NSStringFromClass(PendingContribution.class)
+                                      inManagedObjectContext:context];
+    
+    NSError *error;
+    const auto fetchedObjects = [context executeFetchRequest:fetchRequest error:&error];
+    if (error) {
+      NSLog(@"%@", error);
+      completion(NO);
+      return;
+    }
+    
+    for (PendingContribution *pc in fetchedObjects) {
```

Comment:
if there can be a large number of pc's then we should use `NSBatchDeleteRequest ` here instead. Once done we can merge changes into viewContext.

Reply:
We've opted not to use batch deletions because they aren't supported in non-SQLite persistent storage. We _could_ technically do both, one for the in memory storage and one for the SQLite one... but not sure if it's worth for the low number of pending contributions a user will likely have